modules/images/image-loading-optimization/class-ilo-html-tag-processor.php

```diff
@@ -226,21 +226,30 @@ public function open_tags(): Generator {
 	/**
 	 * Gets breadcrumbs for the current open tag.
 	 *
-	 * Breadcrumbs are constructed to match the format from detect.js.
+	 * A breadcrumb consists of a tag name and its sibling index.
 	 *
-	 * TODO: Consider rather each breadcrumb being a (tag,index) tuple.
-	 *
-	 * @return array<array{tag: string, index: int}> Breadcrumbs.
+	 * @return Generator<array{string, int}> Breadcrumb.
 	 */
-	public function get_breadcrumbs(): array {
-		$breadcrumbs = array();
+	private function get_breadcrumbs(): Generator {
 		foreach ( $this->open_stack_tags as $i => $breadcrumb_tag_name ) {
-			$breadcrumbs[] = array(
-				'tag'   => $breadcrumb_tag_name,
-				'index' => $this->open_stack_indices[ $i ],
-			);
+			yield array( $breadcrumb_tag_name, $this->open_stack_indices[ $i ] );
```

Comment:
Any particular reason for the data type change here? I find an associative array with named keys easier to understand than an indexed array that isn't really for a list of something.

Reply:
Mainly because it makes it more compact and slightly easier to iterate over per below:

```
foreach ( $this->get_breadcrumbs() as list( $tag_name, $index ) ) {
```

Originally too I had envisioned that a breadcrumbs could have other keys, like a class name or maybe some attributes to make it more like a CSS selector. But this didn't end up being the case, as we only ever need the tag name and index. So making it a tuple seems to make sense to me.

Reply:
I realize it's a few fewer lines of code, but it's harder to understand. Keys allow clarifying what something is, while now that's less apparent, particularly when looking at where the function is used and only finding `list`.

I'd prefer to go for clarity over compactness, but not a blocker as long as this function remains purely internal.